File: core/StickersToSpellWord/StickersToSpellWord.java
package com.dbc;

import java.util.Arrays;

public class StickersToSpellWord {
    public int minStickers(String[] stickers, String target) {
        int m = target.length();
        int[] memo = new int[1 << m];
        Arrays.fill(memo, -1);
        memo[0] = 0;
        int res = dp(stickers, target, memo, (1 << m) - 1);
        return res <= m ? res : -1;
    }

    public int dp(String[] stickers, String target, int[] memo, int mask) {
        int m = target.length();
        if (memo[mask] < 0) {
            int res = m + 1;
            for (String sticker : stickers) {
                int left = mask;
                int[] cnt = new int[26];
                for (int i = 0; i < sticker.length(); i++) {
                    cnt[sticker.charAt(i) - 'a']++;
                }
                for (int i = 0; i < target.length(); i++) {
                    char c = target.charAt(i);
                    if (((mask >> i) & 1) == 1 && cnt[c - 'a'] > 0) {
                        cnt[c - 'a']--;
                        left ^= 1 << i;
                    }
                }
                if (left < mask) {
                    res = Math.min(res, dp(stickers, target, memo, left) + 1);
                }
            }
            memo[mask] = res;
        }
        return memo[mask];
    }
}

File: core/StickersToSpellWord/StickersToSpellWord.py
from typing import List
from functools import cache
from collections import Counter


def minStickers(stickers: List[str], target: str) -> int:
    m = len(target)
    @cache
    def dp(mask: int) -> int:
        if mask == 0:
            return 0
        res = m + 1
        for sticker in stickers:
            left = mask
            cnt = Counter(sticker)
            for i, c in enumerate(target):
                if mask >> i & 1 and cnt[c]:
                    cnt[c] -= 1
                    left ^= 1 << i
            if left < mask:
                res = min(res, dp(left) + 1)
        return res
    res = dp((1 << m) - 1)
    return res if res <= m else -1


